### PR TITLE
feat: add ability to pass more arguments through the toggle script

### PR DESCRIPTION
**Pull request by cenk1cenk2**
Thank you sincerely for this, I am trying to build a neovim plugin out of this.

- Adds the ability to pass more arguments to toggle script which will be able to change the socket path and the id format on the fly.

### CODE REVIEW

File: scripts/toggle.sh [hunk 1]
```diff
@@ -49,22 +53,29 @@ cmd=("${@:$OPTIND}")
 
 opened="$(showvariable @__popup_opened)"
 
-if [[ -n "$opened" && ("$opened" = "$name" || -z "$*") ]]; then
+if [[ -n "$opened" ]]; then
```
Comment (loichyan) [inline scripts/toggle.sh:56]: The removed check is necessary: If the specified name doesn't match the currently opened popup, we open a new popup within the current one (i.e. popup-in-popup).

File: scripts/toggle.sh [hunk 2]
```diff
@@ -31,10 +31,14 @@ while getopts :-:BCEb:c:d:e:h:s:S:t:T:w:x:y: OPT; do
...
 
-			  --name <name>     Popup name [Default: "default"]
-			  -[BCE]            Flags passed to display-popup
-			  -[bcdehsStTwxy] <value>
-			                    Options passed to display-popup
+			  --name <name>                   Popup name. [Default: "$DEFAULT_NAME"]
+			  --socket-name <value>           Socket name. [Default: "$DEFAULT_SOCKET_NAME"]
+			  --id-format <value>             Popup ID format. [Default: "$DEFAULT_ID_FORMAT"]
+			  --on-init <hook>                Command to run on popup initialization. [Default: "$DEFAULT_ON_INIT"]
+				--popup-before-open <hook>      Hook to run before opening the popup.
```
Comment (loichyan) [inline scripts/toggle.sh:38]: Here is a nearly invisible formatting issue: this line `--popup-before-open` is preceded by a tab, while the other lines are preceded by two spaces.

Also, this argument should be named `--before-open` to maintain consistency with the others.

File: scripts/toggle.sh [hunk 3]
```diff
@@ -31,10 +31,14 @@ while getopts :-:BCEb:c:d:e:h:s:S:t:T:w:x:y: OPT; do
...
+				--popup-after-close <hook>      Hook to run after closing the popup.
```
Comment (loichyan) [inline scripts/toggle.sh:39]: Ditto.

File: scripts/toggle.sh [hunk 4]
```diff
@@ -7,19 +7,19 @@ source "$CURRENT_DIR/helpers.sh"
...
-declare name popup_args cmd OPT OPTARG OPTIND=1
+declare name socket_name id_format popup_args cmd OPT OPTARG OPTIND=1
 
 while getopts :-:BCEb:c:d:e:h:s:S:t:T:w:x:y: OPT; do
 	if [[ $OPT = '-' ]]; then OPT="${OPTARG%%=*}"; fi
 	case "$OPT" in
 	[BCE]) popup_args+=("-$OPT") ;;
 	[bcdehsStTwxy]) popup_args+=("-$OPT" "$OPTARG") ;;
-	name)
+	name | socket-name | id-format | on-init)
```
Comment (loichyan) [inline scripts/toggle.sh:17]: Missing `--popup-before-open` and `--popup-after-close`.

File: scripts/toggle.sh [hunk 5]
```diff
@@ -49,22 +53,29 @@ cmd=("${@:$OPTIND}")
...
-	eval "tmux -C \; $(showhook @popup-before-open) >/dev/null"
-	tmux popup "${popup_args[@]}" "
+
+	exit 0
+fi
+
+name="${name:-"$DEFAULT_NAME"}"
+socket_name="${socket_name:-$(get_socket_name)}"
+id_format="${id_format:-$(showopt @popup-id-format "$DEFAULT_ID_FORMAT")}"
+on_init="${on_init:-$(showopt @popup-on-init "$DEFAULT_ON_INIT")}"
```
Comment (loichyan) [inline scripts/toggle.sh:67]: For hooks, we should use `showhook`, which performs some necessary processing, such as handling line breaks.

**Review by loichyan — COMMENTED** [review]
Brilliant, thank you very much! However, there are a few minor issues that need to be addressed. I have already made the corrections. If there are no more changes, we can merge this PR :)

**Comment by cenk1cenk2** [comment]
Wow, I was just drafting the idea and you seem to be already on board :D, Thanks @loichyan, I will look at your comments.

File: scripts/toggle.sh (hunk 1 above)
Comment (cenk1cenk2) [reply scripts/toggle.sh]: Oh I see, I was thinking something like neovim opens a popup and I also have a generically set popup in a tmux keybind, I can either use that to close the popup that is opened up by neovim or toggle a new one if that is not open.

So the neovim popup is named differently from the other one, so it was causing this popup in a popup. I just did not think of it as a feature :D

Comment (loichyan) [reply scripts/toggle.sh]: > Oh I see, I was thinking something like neovim opens a popup and I also have a generically set popup in a tmux keybind, I can either use that to close the popup that is opened up by neovim or toggle a new one if that is not open.

I have changed the testing conditions to `[[ -z $name || $name == $opened_name ]]`, which means that an empty `--name` closes the currently opened popup. So, if you have bound a key to `run "#{@popup-toggle} bash"`, pressing that key will open Bash in your main session and will close any open popups when inside them. The user can use it as the primary toggle key.

Personally, I use several different keys to toggle various programs (fish, lazygit, yazi, etc.). I will add some notes for "popup-in-popup" later.

**Comment by cenk1cenk2** [comment]
Thank you for your amazing work on this!

I have also drafted my `neovim` plugin out of this, albeit it still needs some work.

One caveat is I could not make it so for the `neovim` and toggling to disable to popup in popup behavior. I clearly understand that you like this behavior and it is your plugin, so I added something undestructive where if you pass `--single-instance` flag it always closes whatever open first. So it is like a real toggle. I sincerely hope that is acceptable in your eyes.

The other caveat is I know that while creating a `tmux` popup, I know that I can pass an `-e` to pass a new set of environment variables, where I want to pass the `neovim` socket path so that `flatten.nvim` plugin works. However for the life of me I can not get this working, I am thinking that this is related to always creating a new `tmux` session with `tmux -L new ...` it just not behave as described in the documentation.

You can find the draft of the plugin if you are interested in [here](https://github.com/cenk1cenk2/tmux-toggle-popup.nvim), it currently has no proper readme but it should be easy to follow through since it basically wraps the `tmux` commands.

**Comment by loichyan** [comment]
>  So it is like a real toggle. I sincerely hope that is acceptable in your eyes.

Certainly ! :) I was thinking about some ideas like this.

> The other caveat is I know that while creating a `tmux` popup, I know that I can pass an `-e` to pass a new set of environment variables, where I want to pass the `neovim` socket path so that `flatten.nvim` plugin works.

That's intriguing. I will try to work on this later.


> You can find the draft of the plugin if you are interested in [here](https://github.com/cenk1cenk2/tmux-toggle-popup.nvim)

I've had a quick look into your repo, and I am thinking it may not work as you expected (I may not have correctly understood how to use this plugin):

1. The user bound key `<C-T>` to `require("tmux-toggle-popup").toggle()`.
2. They were working in tmux session _A_, and they pressed `<C-T>` to a new popup session _B_.
3. After some work, they were going to close the popup by pressing `<C-T>` again.
4. Popup session _B_ received the key `<C-T>`, while nvim was running in session _A_ and it would not receive `<C-T>`, so popup _B_ would never be closed!

Popups in tmux are little tricky, whenever you open a new one, and move your focus into it, any further input events are sent to the popup instead of your original session. When handling the toggleable popup (let's say _C_), we actually attach a tmux session _C_ in the popup, and when `#{popup-toggle}` is called again inside _C_, we simply detach the session _C_, in order to persist your work in the popup.

Therefore, if the user want to use `<C-T>` to close _B_, they have to create a temporary keybinding inside _B_. That's why I recommend the user to use a primary toggle key.

=====

>  So it is like a real toggle. I sincerely hope that is acceptable in your eyes.

Certainly! :) I was thinking about some ideas like that.

> The other caveat is that while creating a `tmux` popup, I know that I can pass an `-e` to provide a new set of environment variables, where I want to pass the `neovim` socket path so that the `flatten.nvim` plugin works.

That's intriguing. I will try to work on this later.

> You can find the draft of the plugin if you are interested in [here](https://github.com/cenk1cenk2/tmux-toggle-popup.nvim)

I've had a quick look at your repo, and I am thinking it may not work as you expected (I may not have correctly understood how to use your plugin):

1. The user bound the key `<C-T>` to `require("tmux-toggle-popup").toggle()`.
2. They were working in tmux session _A_, and they pressed `<C-T>` to open a new popup session _B_.
3. After some work, they intended to close the popup by pressing `<C-T>` again.
4. Popup session _B_ received the key `<C-T>`, while nvim was running in session _A_ and did not receive `<C-T>`, so popup _B_ would never close!

Popups in tmux are a little tricky. Whenever you open a new one and move your focus into it, any further input events are sent to the popup instead of your original session. When handling the toggleable popup (let's say _C_), we actually attach tmux session _C_ in the popup, and when `#{popup-toggle}` is called again inside _C_, we simply detach session _C_ to persist your work in the popup and allow you to attach to it again.

Therefore, if the user wants to use `<C-T>` to close _B_, they must create a temporary keybinding inside _B_ (this can be accomplished through `--on-init`, but this means they must pass their desired key to `require("tmux-toggle-popup").toggle()`). That's why I recommend that users utilize a primary toggle key.

**Comment by loichyan** [comment]
> One caveat is I could not make it so for the `neovim` and toggling to disable to popup in popup behavior. I clearly understand that you like this behavior and it is your plugin, so I added something undestructive where if you pass `--single-instance` flag it always closes whatever open first. So it is like a real toggle. I sincerely hope that is acceptable in your eyes.

This should be fixed in #6. This issue is basically caused by the same reason as the aforementioned toggle key issue.

**Comment by cenk1cenk2** [comment]
Oh yeah, it is a little bit tricky as you imagined, so it is not a real toggle in this case.

So let us imagine that I have a global toggle key for having a scratch terminal at `C-a C-a`. That uses the `tmux` flow and creates a scratch terminal for me that is shared accross since I have a simpler configuration.

```tmux
set -g @popup-socket-name 'default'
set -g @popup-id-format '#{session_name}/#{@popup_name}'
```

Inside the `neovim`, it is exactly like you said, I can toggle with a `neovim` specific keybind like `<F1>`, however as you said it catches the focus, so the closing is only possible through my global keymap `<C-a><C-a>`. In the sense, it is not a real toggle, but on `<C-a><C-a>` I can access my general scratch buffer and inside the `neovim`, I now have a separate session, specific to `neovim`. The idea is to add more toggles like `<Space>tg` to trigger `lazygit` to mostly mimic what I already have.

The config for neovim is [here](https://github.com/cenk1cenk2/nvim/blob/rolling/lua/ck/plugins/tmux-toggle-popup-nvim.lua) for the plugin albeit it is a bit more abstracted.

**Comment by loichyan** [comment]
> The idea is to add more toggles like `<Space>tg` to trigger `lazygit` to mostly mimic what I already have.

The only solution in my mind is to set temporary keybindings as mentioned in <https://github.com/loichyan/tmux-toggle-popup/pull/5#issuecomment-2415373776>. It seems there aren't any better options :(

AFAIK, tmux doesn't allow binding multiple key sequences, which is unfortunate but simplifies the transpilation from NeoVim keys to tmux keys [^1]. Therefore, you may provide an API like `create_toggle({ key = "<F1>", command = "..." })`, which requires the binding key.

BTW, I would like to change `--single-instance` to `--toggle`, and also document this argument in the README. I'm going to do this later.

[^1]: tmux and NeoVim has different key representions, e.g. `F1` in tmux and `<F1>` in NeoVim. I believe, typically, removing surrounding angle brackets is enough to do so.

**Comment by cenk1cenk2** [comment]
I have finished the readme on my neovim plugin and other than the short coming you are already aware of with the focus it really works like charm in my eyes. Let me see about that whenever the honeymoon phase ends.

But honestly, I can not thank you enough for implementing these features.

**Comment by loichyan** [comment]
> I have finished the readme on my neovim plugin

Ohhh, awesome! But I’ve just changed the argument name. You may update the documentation and the corresponding Lua code later. Sorry for the extra work.

**Pull request merged**